### PR TITLE
Fix Issue 19370 - AA require() and update() can't be used in @safe code

### PR DESCRIPTION
**Pull request by radcapricorn**

### CODE REVIEW

File: src/object.d [hunk 1]
```diff
@@ -3163,7 +3163,18 @@ inout(V) get(K, V)(inout(V[K])* aa, K key, lazy inout(V) defaultValue)
 ref V require(K, V)(ref V[K] aa, K key, lazy V value = V.init)
 {
     bool found;
-    auto p = cast(V*) _aaGetX(cast(void**)&aa, typeid(V[K]), V.sizeof, &key, found);
+    // if key and value are @safe-ly copyable, `require` can infer @safe
+    static if (isSafeCopyable!K && isSafeCopyable!V)
```
Comment (jacob-carlborg) [inline src/object.d]: How does this condition help, why not always use the `@trusted` wrapper?

Comment (radcapricorn) [reply src/object.d]: Maybe that wasn't that clear of an answer: if we just wrap the whole call into a @trusted, it's going to allow un-@safe key postblits. The unittest where S is key will fail.

Comment (jacob-carlborg) [reply src/object.d]: I still don't understand. Your running exactly the same code regardless if it's `isSafeCopyable` or not.

Comment (radcapricorn) [reply src/object.d]: Yes, the code is the same, but the inferred attributes for update()/require() will be different. _aaGetX internally calls the key's postblit. That postblit may be user-defined. If that postblit is `@system`, the whole require()/update() call should be `@system`. But if I were to wrap the call to _aaGetX into a `@trusted` unconditionally, I'd in effect silently promote `@system` user code to `@safe`.

Comment (jacob-carlborg) [reply src/object.d]: Ah, ok, I see. Thanks for explaining. Wasn't every obvious to me at least.

Comment (radcapricorn) [reply src/object.d]: No problem.

Comment (thewilsonator) [reply src/object.d]: This good to go?

Comment (radcapricorn) [reply src/object.d]: By me? Yes :)

Comment (thewilsonator) [reply src/object.d]: Please squash.

File: src/object.d [hunk 2]
```diff
@@ -3205,13 +3219,54 @@ void update(K, V, C, U)(ref V[K] aa, K key, scope C create, scope U update)
...
         *p = update(*p);
 }
 
+unittest
+{
+    static struct S
+    {
+        int x;
+    @nogc nothrow pure:
+        this(this) /* not @safe */ {}
```
Comment (jacob-carlborg) [inline src/object.d]: Perhaps say `@system` here to make it extra clear.

Comment (radcapricorn) [reply src/object.d]: Done.

**Pull request merged**